### ISSUE TITLE
Incorrect error propagation in Bio.File.as_handle

### ISSUE THREAD

**Issue by bow** [biopython/biopython#1544]
### Setup

Hi everyone, I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```
3.5.3 (default, Feb 20 2017, 14:45:39) 
[GCC 6.3.1 20170109]
CPython
Linux-4.14.20-1-lts-x86_64-with-arch
1.71.dev0
```
(this is done on the tip of the current master: 27e345755).

### Expected behaviour

When using `Bio.File.as_handle` in a `with` block, TypeErrors from inside the `with` block should be propagated correctly.

### Actual behaviour

In addition to the actual exception traceback, another internal error is shown:

```
<actual exception traceback>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "as_handle.py", line 11, in <module>
    write_to(target, "foo")
  File "/home/bow/.pyenv/versions/3.5.3/lib/python3.5/contextlib.py", line 78, in __exit__
    raise RuntimeError("generator didn't stop after throw()")
RuntimeError: generator didn't stop after throw()
```

### Steps to reproduce

You can run the following script:

```
#!/usr/bin/env python

from Bio.File import as_handle


with as_handle("test.txt", "w") as target:
    raise TypeError("foo")
```

### Additional Notes

I stumbled on this when trying to add keyword arguments for `SeqIO.*` in [my branch](https://github.com/bow/biopython/tree/enh/seqio-kwargs). It seems to be caused by the `except TypeError` block at the end of the `as_handle` function. [Exceptions inside the `with` block is reraised at the point where `yield` occured inside the decorated function](https://docs.python.org/3.6/library/contextlib.html#contextlib.contextmanager), so in our case this means the TypeError is then re-caught by the final `except` block, which then yields another object. This incorrect, however, and I suppos is the reason the `contextlib` authors added a check on it (which raises the second additional exception).

I have a fix already, [in that branch](https://github.com/bow/biopython/blob/enh/seqio-kwargs/Bio/File.py#L85) ~ but I'd just like another set of eyes to confirm the bug and/or the fix in the branch. It simply ensures that `TypeError`s encountered in the `yield` points are tracked correctly and are re-raised in the last `except` block.

If everything looks ok, I'll open a PR shortly.

**Comment by peterjc** [comment]
Hmm, the current try/except seems to be from https://github.com/biopython/biopython/pull/1476 by  @mruffalo to support ``pathlib.Path`` objects. Perhaps we can refactor that object type checking to avoid the current side effect with a context manager?

**Comment by bow** [comment]
Aha, so then this was related to the fact that in Python 3.4, many of the builtin path-related functions that require `str` objects do not yet work with `Path` objects. IIRC this is not the case anymore in 3.5 onwards.

So we can just check for it in `as_handle` by adding a version check:

```
if sys.version_info[:2] < (3, 5):
    from pathlib import Path
    handleish = str(handleish) if isinstance(handleish, Path) else handleish

<the rest of the code without the try-except block>
```

This seems cleaner than tracking the inner exceptions, which I currently do there.

**Comment by peterjc** [comment]
That does sound cleaner (and easier to remove the hack once we drop Python 3.4)

**Comment by bow** [comment]
Great, I've opened PR #1545.

Small note: the `Path` -> `str` compatibility is actually in 3.6, so we would need to have this until we drop Python 3.5 dependency. I've added a note in the code for this.

**Comment by mruffalo** [comment]
Sorry for the unintended behavior change with `TypeError`s inside calls to `as_handle`.

Tracking the exception seems like a _much_ better solution, though, compared to conditional imports based on Python version and checks for `isinstance(handleish, Path)` (or other things like that). `Path` is not the only object in the Python standard library that can be passed to `open`; `os.DirEntry` also works, and there are third-party replacements for `pathlib` that also support PEP 519. Anyone can define a path-like class by adding a `__fspath__` method, which allows objects of that class to be passed to `open`, and third-party packages like BioPython should not actively prevent this by only allowing a hardcoded set of types.

**Comment by bow** [comment]
@mruffalo Thanks for checking back in ~ and for bringing up `__fspath__`. I suppose there is also an additional way of checking for its presence [as noted by the PEP](https://www.python.org/dev/peps/pep-0519/#backwards-compatibility), which we could do as an `elif` block and would ensure that `handleish` is still a string and would work be compatible with `open`:

```
# Already checked if is Python 3.
if 4 <= sys.version_info.minor <= 5:
    ...
elif sys.version_info.minor > 5:
    handleish = handleish.__fspath__() if hasattr(handleish, "__fspath__") else handleish
```

But at this point, I'm starting to feel the `try-catch` block does feel cleaner. We could do some small changes inside the `try` block so that `inner_exc` is only set at one place, if it helps:

```
inner_exc = False
try:
    if sys.version_info[0] >= 3 and "U" in mode:
        mode = mode.replace("U", "")
    func = codecs.open if "encoding" in kwargs else open
    with func(handleish, mode, **kwargs) as fp:
        try:
            yield fp
        except Exception:
            inner_exc = True
            raise
except TypeError:
    if not inner_exc:
        yield handleish
    else:
        raise
```

**Comment by mruffalo** [comment]
It just occurred to me that something simpler might also work: https://github.com/mruffalo/biopython/commit/bbe71ab1a1ac3eb5fa632258da386aeaa522afe0. I haven't yet run the unit test suite after making that change.

I liked the idea of the EAFP-style operation of "try to open it; if it fails, it must already be a file-like object", but I didn't expect how convoluted it might be to handle errors properly.

**Comment by etal** [comment]
Alternatively, [in pandas](https://github.com/pandas-dev/pandas/blob/2532a49c158193766534aadc345537cdaadc9bc3/pandas/core/dtypes/inference.py#L141
) it looks like the approach is to check if an input supports the minimal interface of a file-like object (has a `read` or `write` attribute and `__iter__`), and otherwise try to open it. This seems consistent with Python's definition of a file-like object in [io.IOBase](https://docs.python.org/3/library/io.html#io.IOBase) and the [glossary](https://docs.python.org/3/glossary.html#term-file-object).

**Comment by bow** [comment]
@mruffalo & @peterjc,  if we are doing the `if hasattr(os, "fspath")` check, doesn't that mean that in Python 3.4 and 3.5, we accept that passing `Path` instances to `as_handle` would result in a `TypeError`? `pathlib.Path` is already in the standard library [since Python 3.4](https://docs.python.org/3/library/pathlib.html#module-pathlib), but PEP 519 and `os.fspath` [did not come until Python 3.6](https://docs.python.org/3/library/os.html#os.fspath).

**Comment by mruffalo** [comment]
Yes, that was the intention behind #1476, as documented in https://github.com/biopython/biopython/pull/1476/commits/e7815fe9aef69aef617fddbeb13a67bdb4a0a1ea in the docstring of `File.as_handle`. My addition to NEWS.rst deserved some clarification after #1476, hence #1488 (specifically https://github.com/biopython/biopython/pull/1488/commits/2e8b3d4dfb27d1ab3f3462723cee608eacefb8b0).

**Comment by bow** [comment]
Ok, the fixes have all been merged in #1555.
